ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -133,107 +222,245 @@ long getMaxBlobSize() {
+    ScanResults currentScanResults = scanResults.get();
+    if (currentScanResults != null && isWithinRange(currentScanResults.containerForecastStartTimeMs,
+        currentScanResults.containerForecastEndTimeMs, referenceTimeInMs)) {
+      retValue = currentScanResults.getValidSizePerContainer(referenceTimeInMs);
+    } else {
+      if (isScanning && isWithinRange(indexScanner.newScanResults.containerForecastStartTimeMs,
+          indexScanner.newScanResults.containerForecastEndTimeMs, referenceTimeInMs)) {
+        scanLock.lock();
+        try {
+          if (isScanning) {
+            if (waitCondition.await(waitTimeoutInSecs, TimeUnit.SECONDS)) {
+              currentScanResults = scanResults.get();
+              if (isWithinRange(currentScanResults.containerForecastStartTimeMs,
+                  currentScanResults.containerForecastEndTimeMs, referenceTimeInMs)) {
+                retValue = currentScanResults.getValidSizePerContainer(referenceTimeInMs);
+              }
+            } else {
+              metrics.blobStoreStatsIndexScannerErrorCount.inc();
+              logger.error("BlobStoreStats index scan took too long to finish for store {}", storeId);
+            }
+          }
+        } catch (InterruptedException e) {
+          metrics.blobStoreStatsIndexScannerErrorCount.inc();
+          throw new IllegalStateException(
+              String.format("Illegal state, wait for scan to complete is interrupted for store %s", storeId), e);
+        } finally {
+          scanLock.unlock();
+        }
+      }
+      if (retValue == null) {
+        retValue = collectValidDataSizeByContainer(referenceTimeInMs);
+      }
+    }
+    return retValue;
+  }
+
+  /**
+   * Function that handles new PUT after a scan to keep the current {@link ScanResults} relevant.
+   * @param putValue the {@link IndexValue} of the new PUT
+   */
+  void handleNewPutEntry(IndexValue putValue) {
+    if (bucketingEnabled && recentEntryQueueEnabled) {
+      recentEntryQueue.offer(new Pair<IndexValue, IndexValue>(putValue, null));
+      metrics.statsRecentEntryQueueSize.update(queueEntryCount.incrementAndGet());
+    }
+  }
+
+  /**
+   * Function that handles new DELETE after a scan to keep the current {@link ScanResults} relevant.
+   * @param deleteValue the {@link IndexValue} of the new DELETE
+   * @param originalPutValue the {@link IndexValue} of the original PUT that is getting deleted
+   */
+  void handleNewDeleteEntry(IndexValue deleteValue, IndexValue originalPutValue) {
+    if (bucketingEnabled && recentEntryQueueEnabled) {
+      recentEntryQueue.offer(new Pair<>(deleteValue, originalPutValue));
+      metrics.statsRecentEntryQueueSize.update(queueEntryCount.incrementAndGet());
+    }
+  }
+
+  /**
+   * Disable this {@link BlobStoreStats} and cancel any ongoing and scheduled index scan.
+   * @throws InterruptedException
+   */
+  @Override
+  public void close() {
+    enabled.set(false);
```

Comment:
using `compareAndSet()` might be better since you need not execute the rest of the statements if this is already `false`, right?